### DISCUSSION TITLE
Using auto anchors decreases mAP and increases loss

### DISCUSSION THREAD

**Discussion by hoseinhamzei** [ultralytics/yolov5#8146] [Q&A]
Hello again and thanks for this great project and your support @glenn-jocher
I've been experimenting with different options and modifications to maximize the model performance for my dataset, I trained yolov5 L on a modified visDrone dataset from scratch which has small objects (my purpose is detecting small cars, etc) both with and without auto anchors while I expected the model to perform way better with auto anchors since the objects in my dataset are much smaller than coco but the model performed worse with auto anchors.
the mAP0.5 with auto anchors after 50 epochs is 0.46 while with --noautoanchors option surprisingly the mAP increased to 49.6 in only 40 epochs! and the box loss decreased as well, it's the same case for mAP0.95

**without auto anchors:**
[noautoanchors wandb report](https://wandb.ai/hosein/yolo5l/reports/yolov5-L-scratch-noautoanchors--VmlldzoyMTMzNzk2?accessToken=0qsjbsg6f61m9txd374lqv9by96a8g05etko0a0mi3atv7zcnov9vmspo6ku9atg)

**with auto anchors:**
[with auto anchor wandb report](https://wandb.ai/hosein/yolo5l/reports/yolov5-L-autoanchors--VmlldzoyMTMzODA5?accessToken=h1flax0zjbhyx6brl3ccgzg2mtncna8681f2g99mnpas3ug7nkhjgnfb6ly8bk0d)

Now my problem is that I want to train yolov5-p2 model(or maybe an even more customized model) for detecting small objects but in the yaml file there are no static anchors and only auto anchor is available in this model and p2 anchors aren't available in the `models/hub/anchors.py` file either. I trained the p2 model with auto anchors and just like normal yolo5L the box loss is the same and mAP while is better, could be much better with static anchors if I had them. does any one have static coco anchors for yolo5 p2 model or any tips for achieving better results with auto anchors?

**Comment by glenn-jocher** [comment]
@hoseinhamzei 👋 Hello! Thanks for asking about model anchors. YOLOv5 🚀 uses a new [Ultralytics](https://ultralytics.com/) algorithm called **AutoAnchor** for anchor verification and generation before training starts.

Autoanchor will analyse your anchors against your dataset and training settings (like `--img-size`), and will adjust your anchors as necessary if it determines the original anchors are a poor fit, or if an **anchor count** was specified in your model.yaml rather than **anchor values**, i.e. 
```yaml
# Specify anchor count (per layer)
anchors: 3

# --OR-- Specify anchor values manually
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
```

When generating new anchors, autoanchor first applies a kmeans function against your dataset labels (scaled to your training `--img-size`), and uses kmeans centroids as initial conditions for a [Genetic Evolution](https://en.wikipedia.org/wiki/Genetic_algorithm) (GE) algorithm. The GE algorithm will evolve all anchors for 1000 generations under default settings, using CIoU loss (same regression loss used during training) combined with [Best Possible Recall](https://en.wikipedia.org/wiki/Precision_and_recall) (BPR) as its fitness function.

Notebook example: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img width="1200" alt="Screenshot 2022-03-24 at 10 58 28" src="https://user-images.githubusercontent.com/26833433/159891298-8b15f80b-f5b0-4fa9-8824-7380bae32a76.png">

**No action is required on your part** to use autoanchor. If you would like to force manual anchors for any reason, you can skip autoanchor with the `--noautoanchor` flag:
```shell
python train.py --noautoanchor
```

For more details on AutoAnchor see:
https://github.com/ultralytics/yolov5/blob/master/utils/autoanchor.py

Good luck 🍀 and let us know if you have any other questions!

**Reply by hoseinhamzei** [reply]
thanks, yes auto anchors is a pretty useful algorithm, but unexpectedly it makes the mAP and loss worse for me and there are no static coco anchors for the yolov5-p2 model do you have the anchors for p2 model or have any idea why auto anchors decreases the performance (mAP and losses)?

**Reply by glenn-jocher** [reply]
@hoseinhamzei results may vary. If you'd like to use fixed anchors follow the instructions above.
